org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/XMLTextDocumentService.java

```diff
@@ -385,6 +412,9 @@ public void didClose(DidCloseTextDocumentParams params) {
 	public CompletableFuture<List<Either<Command, CodeAction>>> codeAction(CodeActionParams params) {
 		return computeDOMAsync(params.getTextDocument(), (cancelChecker, xmlDocument) -> {
 			String uri = params.getTextDocument().getUri();
+
+			sharedSettings.getFormattingSettings().merge(getIndentationSettings(uri).join());

```

Comment:
Using _join_ can be dangerous (it could block the process), I think it should be better to use `thenApply`

Reply:
Okay. Would it be a good idea to do a `CompletableFuture.allOf(...)` here for the DOMAsync and the settings request, or should I do these as a chain of `.thenApply(...)`? I am fairly new to asynchronous programming.